### PR TITLE
Refactor Thumbnail

### PR DESCRIPTION
**Pull request by dump247**
* suspend main
* Show common issue with GlobalScope and try/catch
* Show launch in scope and supervisorScope
* Show usage of async/await

### CODE REVIEW

File: src/main/kotlin/com/valbaca/gotlin/ch8/Thumbnail.kt [hunk 1]
```diff
@@ -14,46 +15,48 @@ import kotlin.time.Duration.Companion.milliseconds
...
-This is unlike Go and obviates the example in the book, which is meant to show that Go will exit the main function
-even if there are open/active goroutines.
 https://github.com/adonovan/gopl.io/blob/master/ch8/thumbnail/thumbnail_test.go#L30
  */
-suspend fun makeThumbnails2(filenames: List<String>) = coroutineScope {
+fun makeThumbnails2(filenames: List<String>) {
     for (f in filenames) {
         try {
-            launch { imageFile(f) } // Unlike go, we don't have to worry about loop variable capture inside launch block
+            GlobalScope.launch { imageFile(f) } // Unlike go, we don't have to worry about loop variable capture inside launch block
```
Comment (dump247) [inline src/main/kotlin/com/valbaca/gotlin/ch8/Thumbnail.kt:59]: Using a different scope, GlobalScope, replicates the issue being shown in go. Using GlobalScope is an easy mistake, so good to point it out here.

File: src/main/kotlin/com/valbaca/gotlin/ch8/Thumbnail.kt [hunk 2]
```diff
@@ -14,46 +15,48 @@ import kotlin.time.Duration.Companion.milliseconds
...
+from imageFile because the block passed to `launch` is executed concurrently.
 
-This is unlike Go and obviates the example in the book, which is meant to show that Go will exit the main function
-even if there are open/active goroutines.
 https://github.com/adonovan/gopl.io/blob/master/ch8/thumbnail/thumbnail_test.go#L30
  */
-suspend fun makeThumbnails2(filenames: List<String>) = coroutineScope {
+fun makeThumbnails2(filenames: List<String>) {
     for (f in filenames) {
         try {
```
Comment (dump247) [inline src/main/kotlin/com/valbaca/gotlin/ch8/Thumbnail.kt:58]: I added to the docs that this try/catch does not catch an exception from `imageFile`

File: src/main/kotlin/com/valbaca/gotlin/ch8/Thumbnail.kt [hunk 3]
```diff
@@ -64,109 +67,55 @@ suspend fun makeThumbnails2(filenames: List<String>) = coroutineScope {
 /*
 See above for why this examples doesn't even make much sense
 https://github.com/adonovan/gopl.io/blob/master/ch8/thumbnail/thumbnail_test.go#L41
+
+The `coroutineScope` block within this method does not exit until all the `launch` jobs
+complete. If any one of the launch jobs throws an exception, the scope will be cancelled, all
+other jobs still in progress will be cancelled and the exception is propagated.
  */
 suspend fun makeThumbnails3(filenames: List<String>) = coroutineScope {
```
Comment (dump247) [inline src/main/kotlin/com/valbaca/gotlin/ch8/Thumbnail.kt:75]: In this case, the channel is unnecessary. The `coroutineScope` block will not exit until all of the `launch` operations completes.

File: src/main/kotlin/com/valbaca/gotlin/ch8/Thumbnail.kt [hunk 4]
```diff
@@ -64,109 +67,55 @@ suspend fun makeThumbnails2(filenames: List<String>) = coroutineScope {
...
-    repeat(filenames.size) {
-        val e = exceptions.receive()
-        if (e != null) {
-            throw e // Unlike go, this will NOT result in a coroutine leak. Per the documentation of `coroutineScope`
-            // However, if we used supervisorScope, we would need to protect against this. See makeThumbnails5
-            /* When any child coroutine in this scope fails, this scope fails and all the rest of the children are
-             cancelled (for a different behavior see supervisorScope). */
-        }
-    }
+suspend fun makeThumbnails4(filenames: List<String>) = supervisorScope {
```
Comment (dump247) [inline src/main/kotlin/com/valbaca/gotlin/ch8/Thumbnail.kt:91]: This is different than the corresponding `go` operation. However, I think replicating the go example in this case will just result in the same thing as makeThumbnails3, which didn't seem useful. So I changed it a bit to provide an example of `supervisorScope`.

File: src/main/kotlin/com/valbaca/gotlin/ch8/Thumbnail.kt [hunk 5]
```diff
@@ -64,109 +67,55 @@ suspend fun makeThumbnails2(filenames: List<String>) = coroutineScope {
...
-                throw e
-            }
-        }
-        jobs.add(job)
-    }
-    launch {
-        for (job in jobs) {
-            job.join() // join on the jobs rather than use a WaitGroup
+    buildList {
+        filenames.consumeEach {
```
Comment (dump247) [inline src/main/kotlin/com/valbaca/gotlin/ch8/Thumbnail.kt:112]: `consumeEach` is the same as `for (it in channel)` with the addition that it cancels the channel when it exits.

File: src/main/kotlin/com/valbaca/gotlin/ch8/Thumbnail.kt (hunk 4 above)
Comment (valbaca) [reply src/main/kotlin/com/valbaca/gotlin/ch8/Thumbnail.kt:91]: Thanks, that's exactly the issue I was running into while writing this (and others). Sometimes the Go examples don't translate over, or don't translate over in any meaningful way. Then there are cases where Kotlin has a different set of issues to lookout for.

Overall, I think one thing I'm ruminating on is how much do I try to:

1.  keep to the original Go example (which sometimes doesn't make any sense in the Kotlin world)
2. show *some* Kotlin way to do it, generally sub-optimal
3. show the *good* Kotlin idiomatic way to do it

Even things as minor as looping over a list. In Go there is literally just a for-loop. No map or filter. Whereas in Kotlin you do **and** we generally prefer immutability. 

Then there's also the fact that many of the Go examples are actually intentionally showing a broken way of doing something for the purpose of then later showing how to do it right (but "right" in Go, not Kotlin).

We may have to end up doing something similar:
- What's the Go -> Kotlin direct translation
- What's the proper Kotlin approach

Comment (dump247) [reply src/main/kotlin/com/valbaca/gotlin/ch8/Thumbnail.kt:91]: It's a tough comparison. The go examples are composing primitives available in the go runtime to coordinate concurrent operations in ways that the Kotlin coroutine system was designed to coordinate.

Anyway, I really appreciate the effort and let me know if there is anything else I can do to help. Feel free to reject my PRs or take part of them or whatever you want to do.

Comment (valbaca) [reply src/main/kotlin/com/valbaca/gotlin/ch8/Thumbnail.kt:91]: I greatly appreciate your PRs. They're incredibly helpful and I'm learning a lot (which was a primary point of this whole exercise). 

I'm just thinking out loud about what kind of examples I want to show.

**Review by valbaca — APPROVED** [review]
Fantastic. Thank you! This was really educational.

**Pull request merged**